src/test/fuzz/crypto_diff_fuzz_chacha20.cpp

```diff
@@ -0,0 +1,330 @@
+                    ctarget[i] = c[i];
+            }
+            x->input[12] = j12;
+            x->input[13] = j13;
+            return;
+        }
+        bytes -= 64;
+        c += 64;
+        m += 64;
+    }
+}
+
+void ECRYPT_keystream_bytes(ECRYPT_ctx* x, u8* stream, u32 bytes)
+{
+    u32 i;
+    for (i = 0; i < bytes; ++i)
+        stream[i] = 0;
+    ECRYPT_encrypt_bytes(x, stream, stream, bytes);
+}
+
+FUZZ_TARGET(crypto_diff_fuzz_chacha20)
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+
+    ChaCha20 chacha20;
+    ECRYPT_ctx ctx;
+    // D. J. Bernstein doesn't initialise ctx to 0 while Bitcoin Core initialises chacha20 to 0 in the constructor
+    for (int i = 0; i < 16; i++) {
+        ctx.input[i] = 0;
+    }
+
+    if (fuzzed_data_provider.ConsumeBool()) {
+        const std::vector<unsigned char> key = ConsumeFixedLengthByteVector(fuzzed_data_provider, fuzzed_data_provider.ConsumeIntegralInRange<size_t>(16, 32));
+        chacha20 = ChaCha20{key.data(), key.size()};
+        ECRYPT_keysetup(&ctx, key.data(), key.size() * 8, 0);
+        // ECRYPT_keysetup() doesn't set the counter and nonce to 0 while SetKey() does
+        uint8_t iv[8] = {0, 0, 0, 0, 0, 0, 0, 0};
+        ECRYPT_ivsetup(&ctx, iv);
+    }
+
+    LIMITED_WHILE (fuzzed_data_provider.ConsumeBool(), 3000) {
+        CallOneOf(
+            fuzzed_data_provider,
+            [&] {
+                const std::vector<unsigned char> key = ConsumeFixedLengthByteVector(fuzzed_data_provider, fuzzed_data_provider.ConsumeIntegralInRange<size_t>(16, 32));
+                chacha20.SetKey(key.data(), key.size());
+                ECRYPT_keysetup(&ctx, key.data(), key.size() * 8, 0);
+                // ECRYPT_keysetup() doesn't set the counter and nonce to 0 while SetKey() does
+                uint8_t iv[8] = {0, 0, 0, 0, 0, 0, 0, 0};
+                ECRYPT_ivsetup(&ctx, iv);
+            },
+            [&] {
+                uint64_t iv = fuzzed_data_provider.ConsumeIntegral<uint64_t>();
+                chacha20.SetIV(iv);
+                ctx.input[14] = iv;
+                ctx.input[15] = iv >> 32;
+            },
+            [&] {
+                uint64_t counter = fuzzed_data_provider.ConsumeIntegral<uint64_t>();
+                chacha20.Seek(counter);
+                ctx.input[12] = counter;
+                ctx.input[13] = counter >> 32;
+            },
+            [&] {
+                uint32_t integralInRange = fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, 4096);
+                std::vector<uint8_t> output(integralInRange);
+                chacha20.Keystream(output.data(), output.size());
+                std::vector<uint8_t> djb_output(integralInRange);
+                ECRYPT_keystream_bytes(&ctx, djb_output.data(), djb_output.size());
+                if (output.data() != NULL && djb_output.data() != NULL) {
+                    assert(memcmp(output.data(), djb_output.data(), integralInRange) == 0);
```

Comment:
any reason to use a low level function, when https://en.cppreference.com/w/cpp/container/vector/operator_cmp could be used?

Reply:
I went with this because [some of the crypto unit tests](https://github.com/bitcoin/bitcoin/blob/4ad59042b359f473d5888ecee0c9288dcf98f1c9/src/test/crypto_tests.cpp#L636) used memcmp. I feel you're right though and memcmp doesn't really provide any added advantage.